File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AutoTPA.java
package me.zeroeightsix.kami.module.modules.chat;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.Friends;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.network.play.server.SPacketChat;

import static me.zeroeightsix.kami.util.MessageSendHelper.sendServerMessage;

/*
 * By Katatje 8 Dec 2019
 * Updated by S-B99 on 12/04/20
 */
@Module.Info(
        name = "AutoTPA",
        description = "Automatically decline or accept TPA requests",
        category = Module.Category.CHAT
)
public class AutoTPA extends Module {
    private Setting<Boolean> friends = register(Settings.b("Always accept friends", true));
    private Setting<mode> mod = register(Settings.e("Response", mode.DENY));

    @EventHandler
    public Listener<PacketEvent.Receive> receiveListener = new Listener<>(event -> {
        if (event.getPacket() instanceof SPacketChat && ((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().contains(" has requested to teleport to you.")) {
            /* I tested that getting the first word is compatible with chat timestamp, and it as, as this is Receive and chat timestamp is after Receive */
            String firstWord = ((SPacketChat) event.getPacket()).getChatComponent().getUnformattedText().split("\\s+")[0];
            if (friends.getValue() && Friends.isFriend(firstWord)) {
                sendServerMessage("/tpaccept");
                return;
            }
            switch (mod.getValue()) {
                case ACCEPT:
                    sendServerMessage("/tpaccept");
                    break;
                case DENY:
                    sendServerMessage("/tpdeny");
                    break;
            }
        }
    });

    public enum mode {
        ACCEPT, DENY
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AutoTPA.kt
package me.zeroeightsix.kami.module.modules.chat

import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.manager.managers.FriendManager
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.text.MessageDetection
import me.zeroeightsix.kami.util.text.MessageSendHelper.sendServerMessage
import net.minecraft.network.play.server.SPacketChat
import org.kamiblue.event.listener.listener

@Module.Info(
    name = "AutoTPA",
    description = "Automatically accept or decline /TPAs",
    category = Module.Category.CHAT
)
object AutoTPA : Module() {
    private val friends = setting("AlwaysAcceptFriends", true)
    private val mode = setting("Response", Mode.DENY)

    private enum class Mode {
        ACCEPT, DENY
    }

    init {
        listener<PacketEvent.Receive> {
            if (it.packet !is SPacketChat || MessageDetection.Other.TPA_REQUEST detectNot it.packet.chatComponent.unformattedText) return@listener

            /* I tested that getting the first word is compatible with chat timestamp, and it as, as this is Receive and chat timestamp is after Receive */
            val name = it.packet.chatComponent.unformattedText.split(" ")[0]

            when (mode.value) {
                Mode.ACCEPT -> sendServerMessage("/tpaccept $name")
                Mode.DENY -> {
                    if (friends.value && FriendManager.isFriend(name)) {
                        sendServerMessage("/tpaccept $name")
                    } else {
                        sendServerMessage("/tpdeny $name")
                    }
                }
            }
        }
    }
}